### PR TITLE
Expand error when discovery document request fails

### PR DESCRIPTION
**Pull request by annawinkler**
Expand the error displayed when a user encounters the discovery document request failure from a TFC CLI run. 

Initially, I was going to look for the error `Client.Timeout exceeded while awaiting headers`, but in my local tests, I was also encountering two other errors:
* `context deadline exceeded`
* `TLS handshake timeout`

So, I decided to look for the error from terraform-svchost: `Failed to request discovery document`

Here is a screenshot of the updated error message:
Screen Shot 2022-12-14 at 2 39 50 PM

Addresses #22774

### Related PRs
* https://github.com/hashicorp/terraform-svchost/pull/10

## Target Release

1.4.x

### ENHANCEMENTS

- Expand the error when the discovery document request fails in Terraform Cloud to indicate that the error is sometimes caused by network connection problems.

### CODE REVIEW

File: internal/cloud/backend.go
Comment (brandonc) [inline internal/cloud/backend.go:421]: I see a couple of nagging issues with this code, although I think it works fine:

1. There's no regex control characters in the pattern, and regex patterns in go typically use backtick quotes to avoid the need for double-escaping control characters (It's usually a linter rule, not sure if that's the case here). I would probably just use [strings.HasPrefix](https://pkg.go.dev/strings#HasPrefix) for this situation to avoid the issue altogether.
2. The error is wrapped, but in a couple of unconventional ways: The wrapped error details are shown first, followed by the added context. Usually it's reversed.
3. The error should be wrapped using %w, which could help upstream callers to [unwrap](https://go.dev/blog/go1.13-errors) to parse the error hierarchy if they want.

```suggestion
		if strings.HasPrefix(err.Error(), "Failed to request discovery document") {
			return nil, fmt.Errorf("A network issue prevented cloud configuration: %w", err.Error())
		}
```

**Review by brandonc — CHANGES_REQUESTED** [review]

File: internal/cloud/backend.go
Comment (annawinkler) [inline internal/cloud/backend.go:416]: Trying to add retry logic here, but it's not retrying. Output from a local run:

```
$ TF_LOG=debug /Users/awinkler/.asdf/installs/golang/1.19.4/packages/bin/terraform apply
2023-01-10T19:15:11.644-0700 [INFO]  Terraform version: 1.4.0 dev
2023-01-10T19:15:11.644-0700 [DEBUG] using github.com/hashicorp/go-tfe v1.12.0
2023-01-10T19:15:11.644-0700 [DEBUG] using github.com/hashicorp/hcl/v2 v2.15.0
2023-01-10T19:15:11.644-0700 [DEBUG] using github.com/hashicorp/terraform-config-inspect v0.0.0-20210209133302-4fd17a0faac2
2023-01-10T19:15:11.644-0700 [DEBUG] using /Users/awinkler/dev/terraform-svchost (devel)
2023-01-10T19:15:11.644-0700 [DEBUG] using github.com/zclconf/go-cty v1.12.1
2023-01-10T19:15:11.644-0700 [INFO]  Go runtime version: go1.19.4
2023-01-10T19:15:11.644-0700 [INFO]  CLI args: []string{"/Users/awinkler/.asdf/installs/golang/1.19.4/packages/bin/terraform", "apply"}
2023-01-10T19:15:11.644-0700 [DEBUG] Attempting to open CLI config file: /Users/awinkler/.terraformrc
2023-01-10T19:15:11.645-0700 [INFO]  Loading CLI configuration from /Users/awinkler/.terraformrc
2023-01-10T19:15:11.645-0700 [INFO]  Loading CLI configuration from /Users/awinkler/.terraform.d/credentials.tfrc.json
2023-01-10T19:15:11.645-0700 [DEBUG] Explicit provider installation configuration is set
2023-01-10T19:15:11.645-0700 [INFO]  CLI command args: []string{"apply"}
2023-01-10T19:15:11.646-0700 [DEBUG] Before calling Discover...
2023-01-10T19:15:11.646-0700 [DEBUG] Service discovery for app.terraform.io at https://app.terraform.io/.well-known/terraform.json
╷
│ Error: Failed to request discovery document: Get "https://app.terraform.io/.well-known/terraform.json": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

Comment (annawinkler) [reply internal/cloud/backend.go:416]: Looked at this with Brandon and saw that the client is being set in the svchost library, so rethinking the problem.

**Review by uturunku1 — COMMENTED** [review]
Hey Anna! I see that this PR is still needs to merge and pull in the latest changes for terraform-svchost. I am happy review and approve after that. Unless, is there any other changes blocking this PR?

**Comment by annawinkler** [comment]
Thanks @uturunku1 ! I've released svchost with the error type and updated the dependencies in this PR, so I think I'm ready for a re-review.

File: internal/cloud/backend.go
Comment (brandonc) [inline internal/cloud/backend.go:425]: I see a couple of code issues (no need for locals, semicolon typo). Should we add the "cloud configuration" context to the other types of errors? Or do we want to lead with "service discovery" errors in that case?

```suggestion
		if errors.Is(err, disco.ErrServiceDiscoveryNetworkRequest) {
			return nil, fmt.Errorf("a network issue prevented cloud configuration: %w", err)
		default:
			return nil, fmt.Errorf("cloud configuration failed: %w", err)
		}
```

**Review by brandonc — APPROVED** [review]

**Pull request merged**
